platform/darwin/src/MGLOfflineStorage.mm

```diff
@@ -357,6 +360,31 @@ - (void)addPackForRegion:(id <MGLOfflineRegion>)region withContext:(NSData *)con
         [[strongSelf mutableArrayValueForKey:@"packs"] addObject:pack];
         if (completion) {
             completion(pack, error);
+            
+            MGLTilePyramidOfflineRegion *tileRegion = MGL_OBJC_DYNAMIC_CAST(region, MGLTilePyramidOfflineRegion);
+            MGLShapeOfflineRegion *shapeRegion = MGL_OBJC_DYNAMIC_CAST(region, MGLShapeOfflineRegion);
+            
+            NSDictionary *attributes;
+            
+            if ([region isKindOfClass:[MGLTilePyramidOfflineRegion class]]) {
+                attributes = @{
+                  @"event": MMEventTypeOfflineDownloadStart,
+                  @"shapeForOfflineRegion": @"tileregion",
+                  @"minZoom": [NSNumber numberWithDouble:tileRegion.minimumZoomLevel],
+                  @"maxZoom": [NSNumber numberWithDouble:tileRegion.maximumZoomLevel],
+                  @"styleURL": region.styleURL
+                  };
+            } else {
```

Comment:
Similarly here, you should check `if (shapeRegion) ...`

However, I think it would be better if we added a readonly property (or method) to `MGLOfflineRegion_Private`, along the lines of:

`@property (nonatomic, readonly) NSDictionary *eventAttributes;`

which both `MGLTilePyramidOfflineRegion` and `MGLShapeOfflineRegion` implement, e.g (tweaking your above attribute dictionary):

```objc
-(NSDictionary *)eventAttributes {
    return @{
        @"shapeForOfflineRegion": @"tileregion",
        @"minZoom": @(self.minimumZoomLevel),
        @"maxZoom": @(self.maximumZoomLevel),
        @"styleURL": self.styleURL // assuming non-null
    };
}
```

Then in this code you'd check for conformance:

```objc
NSMutableDictionary *eventAttributes = [NSMutableDictionary dictionaryWithObject:MMEventTypeOfflineDownloadStart forKey:@"event"];

// Add region specific attributes
if ([region conformsToProtocol:@protocol(MGLOfflineRegion_Private)]) {
    NSDictionary *regionAttributes = ((id<MGLOfflineRegion_Private>)region).eventAttributes; 
    [eventAttributes addEntriesFromDictionary:regionAttributes];
}

[MGLMapboxEvents pushEvent:MMEventTypeOfflineDownloadStart withAttributes:attributes];
```

Reply:
Longer term, we should consider whether having the `pushEvent:withAttributes` call here<sup>1</sup> is the right thing to do. Instead perhaps we delegate or broadcast a notification, that *something* can listen for and it's that object's responsibility to push the event. But this particular topic shouldn't hold up this PR.

---
<sup>1</sup> This also goes for other places we're currently calling `pushEvent...` /cc @fabian-guerra